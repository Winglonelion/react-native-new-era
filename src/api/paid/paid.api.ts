import axios from 'axios';
import {
  PAID_HISTORY_OVERVIEW_ENDPOINT,
  PAID_OVERVIEW_ENDPOINT,
} from '../api.endpoint';
import {
  GetPaidHistoryOverViewParameters,
  GetPaidHistoryOverViewResponse,
  GetPaidOverViewParameters,
  GetPaidOverViewResponse,
} from './paid.api.types';

export function getPaidOverViewAPI({ userId }: GetPaidOverViewParameters) {
  return axios.post<GetPaidOverViewResponse>(PAID_OVERVIEW_ENDPOINT, {
    user_id: userId,
  });
}

export function getPaidHistoryOverViewAPI({
  userId,
}: GetPaidHistoryOverViewParameters) {
  return axios.post<GetPaidHistoryOverViewResponse>(
    PAID_HISTORY_OVERVIEW_ENDPOINT,
    {
      user_id: userId,
    },
  );
}
