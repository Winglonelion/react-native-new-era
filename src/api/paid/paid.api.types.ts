export type GetPaidOverViewParameters = {
  userId: string;
};

export type GetPaidOverViewResponse = {
  take_home: number;
  gross: number;
  hours: number;
  rate: number;
  regular: number;
  over_time: number;
  detail: { type: string; value: number }[];
};

export type GetPaidHistoryOverViewParameters = {
  userId: string;
};

export type GetPaidHistoryOverViewResponse = {
  date: string;
  take_home: number;
  gross: number;
  hours: number;
}[];
