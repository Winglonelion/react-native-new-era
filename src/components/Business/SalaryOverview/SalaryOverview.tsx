import React, { memo } from 'react';
import { StyleSheet, View } from 'react-native';
import { capitalize } from 'lodash';

import Colors from 'theme/colors';
import CommonStyles from 'theme/CommonStyles';
import { mapPieColor } from 'utils/business/chart';
import { GetPaidOverViewResponse } from 'api/paid/paid.api.types';

import TitleText from 'components/Base/Text/TitleText';
import PieChart from 'components/Common/Chart/PieChart';
import MoneyValue from 'components/Base/Text/MoneyValue/MoneyValue';
import ContentLoaderLine from 'components/Base/ContentLoader/BuildIn/ContentLoaderLine';
import ContentLoaderCircle from 'components/Base/ContentLoader/BuildIn/ContentLoaderCircle';
import messages from './SalaryOverview.messages';

export interface PropTypes {
  overviewData?: GetPaidOverViewResponse;
  loading?: boolean;
}

const SalaryOverview: React.FC<PropTypes> = memo(
  ({ overviewData, loading }) => {
    const data = rebuildData(overviewData);
    return (
      <View style={styles.contentRow}>
        <View style={CommonStyles.flex1}>
          {/* take home */}
          <View style={styles.dataRow}>
            <TitleText level="h5" color={Colors.lightBrown}>
              {messages.take_home}
            </TitleText>
            <ContentLoaderLine ready={!loading} width={150} height={30}>
              <TitleText level="h1">
                <MoneyValue value={data?.take_home} />
              </TitleText>
            </ContentLoaderLine>
          </View>
          {/* gross */}
          <View style={styles.dataRow}>
            <TitleText level="h5" color={Colors.lightBrown}>
              {messages.gross}
            </TitleText>
            <ContentLoaderLine ready={!loading} width={150} height={22}>
              <TitleText level="h3">
                <MoneyValue value={data?.gross} />
              </TitleText>
            </ContentLoaderLine>
          </View>
          <View style={[styles.dataRow, CommonStyles.row]}>
            <View style={CommonStyles.flex1}>
              {/* hours */}
              <TitleText level="h5" color={Colors.lightBrown}>
                {messages.hours}
              </TitleText>
              <ContentLoaderLine ready={!loading} width={76} height={22}>
                <TitleText level="h3">{data?.hours}</TitleText>
              </ContentLoaderLine>
            </View>
            <View style={CommonStyles.flex1}>
              {/* rate */}
              <TitleText level="h5" color={Colors.lightBrown}>
                {messages.rate}
              </TitleText>
              <ContentLoaderLine ready={!loading} width={76} height={22}>
                <TitleText level="h3">
                  <MoneyValue value={data?.rate} />
                </TitleText>
              </ContentLoaderLine>
            </View>
          </View>
        </View>
        <View style={[CommonStyles.flex1, styles.chartBox]}>
          <ContentLoaderCircle
            ready={!loading}
            size={150}
            width={150}
            height={150}>
            <PieChart
              style={styles.chart}
              outerRadius={'100%'}
              innerRadius={'65%'}
              padAngle={0}
              startAngle={-20}
              data={data?.dataChart}
            />
          </ContentLoaderCircle>
        </View>
      </View>
    );
  },
);

function rebuildData(data?: GetPaidOverViewResponse) {
  if (!data) return undefined;
  return {
    take_home: data.take_home,
    gross: data.gross,
    hours: data.hours,
    rate: data.rate,
    dataChart: data.detail.map(({ type, value }) => ({
      key: capitalize(type),
      value,
      svg: { fill: mapPieColor(type) },
    })),
  };
}

const styles = StyleSheet.create({
  contentRow: {
    flexDirection: 'row',
    paddingHorizontal: 6,
  },
  dataRow: {
    marginBottom: 8,
  },
  chartBox: {
    justifyContent: 'center',
  },
  chart: {
    height: 150,
  },
});

export default SalaryOverview;
