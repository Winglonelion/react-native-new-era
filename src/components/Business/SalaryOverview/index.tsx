export { default as default } from './SalaryOverview';
