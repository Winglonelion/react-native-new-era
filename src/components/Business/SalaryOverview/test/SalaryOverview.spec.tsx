import React from 'react';
import { shallow } from 'enzyme';
import SalaryOverview, { PropTypes } from '../SalaryOverview';
import { paid_overview } from 'mock/networking/paid.mock';

const setup = (propOverrides: PropTypes = {} as PropTypes) => {
  const props = { ...propOverrides };
  const wrapper = shallow(<SalaryOverview {...props} />);
  return {
    props,
    wrapper,
  };
};

describe('SalaryOverview check snapshot', () => {
  test('should match snapshot with props', () => {
    const { wrapper } = setup({
      overviewData: paid_overview,
      loading: false,
    });
    expect(wrapper).toMatchSnapshot();
  });
  test('should match snapshot without props and loading', () => {
    const { wrapper } = setup({
      loading: true,
    });
    expect(wrapper).toMatchSnapshot();
  });
});
