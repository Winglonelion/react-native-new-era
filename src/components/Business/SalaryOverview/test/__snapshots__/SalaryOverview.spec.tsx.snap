// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalaryOverview check snapshot should match snapshot with props 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "paddingHorizontal": 6,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "marginBottom": 8,
        }
      }
    >
      <Memo()
        color="#9C9079"
        level="h5"
      >
        Take Home
      </Memo()>
      <Memo()
        height={30}
        ready={true}
        width={150}
      >
        <Memo()
          level="h1"
        >
          <Memo()
            value={1000}
          />
        </Memo()>
      </Memo()>
    </View>
    <View
      style={
        Object {
          "marginBottom": 8,
        }
      }
    >
      <Memo()
        color="#9C9079"
        level="h5"
      >
        Gross
      </Memo()>
      <Memo()
        height={22}
        ready={true}
        width={150}
      >
        <Memo()
          level="h3"
        >
          <Memo()
            value={1800}
          />
        </Memo()>
      </Memo()>
    </View>
    <View
      style={
        Array [
          Object {
            "marginBottom": 8,
          },
          Object {
            "flexDirection": "row",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Memo()
          color="#9C9079"
          level="h5"
        >
          Hours
        </Memo()>
        <Memo()
          height={22}
          ready={true}
          width={76}
        >
          <Memo()
            level="h3"
          >
            80
          </Memo()>
        </Memo()>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Memo()
          color="#9C9079"
          level="h5"
        >
          Rate
        </Memo()>
        <Memo()
          height={22}
          ready={true}
          width={76}
        >
          <Memo()
            level="h3"
          >
            <Memo()
              value={30}
            />
          </Memo()>
        </Memo()>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "justifyContent": "center",
        },
      ]
    }
  >
    <Memo()
      height={150}
      ready={true}
      size={150}
      width={150}
    >
      <PieChart
        data={
          Array [
            Object {
              "key": "Taxes",
              "svg": Object {
                "fill": "#77569E",
              },
              "value": 357.74,
            },
            Object {
              "key": "Benefits",
              "svg": Object {
                "fill": "#4FA5AD",
              },
              "value": 73,
            },
            Object {
              "key": "Retirements",
              "svg": Object {
                "fill": "#E28080",
              },
              "value": 40,
            },
            Object {
              "key": "Other",
              "svg": Object {
                "fill": "#C0CAB8",
              },
              "value": 40,
            },
            Object {
              "key": "Take_home",
              "svg": Object {
                "fill": "#FF894D",
              },
              "value": 1000,
            },
          ]
        }
        endAngle={6.283185307179586}
        height={100}
        innerRadius="65%"
        labelRadius="50%"
        outerRadius="100%"
        padAngle={0}
        sort={[Function]}
        startAngle={-20}
        style={
          Object {
            "height": 150,
          }
        }
        valueAccessor={[Function]}
        width={100}
      />
    </Memo()>
  </View>
</View>
`;

exports[`SalaryOverview check snapshot should match snapshot without props and loading 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
      "paddingHorizontal": 6,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "marginBottom": 8,
        }
      }
    >
      <Memo()
        color="#9C9079"
        level="h5"
      >
        Take Home
      </Memo()>
      <Memo()
        height={30}
        ready={false}
        width={150}
      >
        <Memo()
          level="h1"
        >
          <Memo() />
        </Memo()>
      </Memo()>
    </View>
    <View
      style={
        Object {
          "marginBottom": 8,
        }
      }
    >
      <Memo()
        color="#9C9079"
        level="h5"
      >
        Gross
      </Memo()>
      <Memo()
        height={22}
        ready={false}
        width={150}
      >
        <Memo()
          level="h3"
        >
          <Memo() />
        </Memo()>
      </Memo()>
    </View>
    <View
      style={
        Array [
          Object {
            "marginBottom": 8,
          },
          Object {
            "flexDirection": "row",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Memo()
          color="#9C9079"
          level="h5"
        >
          Hours
        </Memo()>
        <Memo()
          height={22}
          ready={false}
          width={76}
        >
          <Memo()
            level="h3"
          />
        </Memo()>
      </View>
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Memo()
          color="#9C9079"
          level="h5"
        >
          Rate
        </Memo()>
        <Memo()
          height={22}
          ready={false}
          width={76}
        >
          <Memo()
            level="h3"
          >
            <Memo() />
          </Memo()>
        </Memo()>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "justifyContent": "center",
        },
      ]
    }
  >
    <Memo()
      height={150}
      ready={false}
      size={150}
      width={150}
    >
      <PieChart
        endAngle={6.283185307179586}
        height={100}
        innerRadius="65%"
        labelRadius="50%"
        outerRadius="100%"
        padAngle={0}
        sort={[Function]}
        startAngle={-20}
        style={
          Object {
            "height": 150,
          }
        }
        valueAccessor={[Function]}
        width={100}
      />
    </Memo()>
  </View>
</View>
`;
